lib/connection/apm.js

```diff
@@ -0,0 +1,221 @@
+const generateConnectionId = pool => `${pool.options.host}:${pool.options.port}`;
+const maybeRedact = (commandName, result) =>
+  SENSITIVE_COMMANDS.indexOf(commandName) !== -1 ? {} : result;
+
+const LEGACY_FIND_QUERY_MAP = {
+  $query: 'filter',
+  $orderby: 'sort',
+  $hint: 'hint',
+  $comment: 'comment',
+  $maxScan: 'maxScan',
+  $max: 'max',
+  $min: 'min',
+  $returnKey: 'returnKey',
+  $showDiskLoc: 'showRecordId',
+  $maxTimeMS: 'maxTimeMS',
+  $snapshot: 'snapshot'
+};
+
+const LEGACY_FIND_OPTIONS_MAP = {
+  numberToSkip: 'skip',
+  numberToReturn: 'batchSize',
+  returnFieldsSelector: 'projection'
+};
+
+const OP_QUERY_KEYS = [
+  'tailable',
+  'oplogReplay',
+  'noCursorTimeout',
+  'awaitData',
+  'partial',
+  'exhaust'
+];
+
+/**
+ * Extract the actual command from the query, possibly upconverting if it's a legacy
+ * format
+ *
+ * @param {Object} command the command
+ */
+const extractCommand = command => {
+  if (command instanceof GetMore) {
+    return {
+      getMore: command.cursorId,
+      collection: command.ns.split('.')[1],
+      batchSize: command.numberToReturn
+    };
+  }
+
+  if (command instanceof KillCursor) {
+    return {
+      killCursors: command.ns.split('.')[1],
+      cursors: command.cursorIds
+    };
+  }
+
+  if (command.query && typeof command.query.$query !== 'undefined') {
+    // upconvert legacy find command
+    const result = { find: command.ns.split('.')[1] };
+    Object.keys(LEGACY_FIND_QUERY_MAP).forEach(key => {
+      if (typeof command.query[key] !== 'undefined')
+        result[LEGACY_FIND_QUERY_MAP[key]] = command.query[key];
+    });
+
+    Object.keys(LEGACY_FIND_OPTIONS_MAP).forEach(key => {
+      if (typeof command.options[key] !== 'undefined')
+        result[LEGACY_FIND_OPTIONS_MAP[key]] = command.options[key];
+    });
+
+    OP_QUERY_KEYS.forEach(key => {
+      if (command[key]) result[key] = command[key];
+    });
+
+    if (typeof command.pre32Limit !== 'undefined') {
+      result.limit = command.pre32Limit;
+    }
+
+    if (command.query.$explain) {
+      return { explain: result };
+    }
+
+    return result;
+  }
+
+  return command.query ? command.query : command;
+};
+
+const extractReply = (command, reply) => {
+  if (command instanceof GetMore) {
+    return {
+      ok: 1,
+      cursor: {
+        id: reply.message.cursorId,
+        ns: command.ns,
+        nextBatch: reply.message.documents
+      }
+    };
+  }
+
+  if (command instanceof KillCursor) {
+    return {
+      ok: 1,
+      cursorsUnknown: command.cursorIds
+    };
+  }
+
+  // is this a legacy find command?
+  if (command.query && typeof command.query.$query !== 'undefined') {
+    return {
+      ok: 1,
+      cursor: {
+        id: reply.message.cursorId,
+        ns: command.ns,
+        firstBatch: reply.message.documents
+      }
+    };
+  }
+
+  return reply.result;
+};
+
+/** An event indicating the start of a given command */
+class CommandStartedEvent {
```

Comment:
`CommandStartedEvent`, `CommandSucceededEvent` and `CommandFailedEvent` seem to have a lot in common. Can we make a common base class and/or use composition to reduce the duplicate logic?

Reply:
Additionally, do these have to be classes? All they have is constructors. It seems like we could instead have factory functions that return POJOs.

Reply:
The original implementation returned POJOs, but I preferred them being well named/defined types. Do you have a case for them being POJOs?

As for duplication, I suppose we could make a common `CommandMonitoringEvent`, but I'm not sure we gain all that much here. I like that I can look at each of these definitions and very clearly see what they doing. I am generally on your side of the fence on issues like these, but when it comes to types like events or errors I prefer serious verbosity (for readabilities sake)